js/src/common/states/PaginatedListState.ts

```diff
@@ -0,0 +1,196 @@
+import Model from '../Model';
+
+export interface Page<TModel> {
+  number: number;
+  items: TModel[];
+}
+
+export interface PaginationLocation {
+  page: number;
+  startIndex?: number;
+  endIndex?: number;
+}
+
+export default abstract class PaginatedListState<T extends Model> {
+  protected location!: PaginationLocation;
+  protected offset: number;
+
+  protected pages: Page<T>[] = [];
+  protected params: any = {};
+
+  protected initialLoading: boolean = false;
+  protected loadingPrev: boolean = false;
+  protected loadingNext: boolean = false;
+
+  hasPrev: boolean = false;
+  hasNext: boolean = false;
+
+  protected constructor(offset: number = 20) {
```

Comment:
Maybe we should try to somewhat mirror https://github.com/flarum/core/blob/c119731e652fcfb96e1bb410e0550c6fbb3dc445/src/Filter/AbstractFilterer.php#L48-L48 (accept params, page size, and initial offset)?

Reply:
Would we use page # instead of initial offset?

Reply:
I think page number and page size would make sense here, since that's what's in the URL. We can add an additional offset later I think